mempool / rpc: followup to getprioritisedtransactions and delete a mapDeltas entry when delta==0

kevkevinpal opened the pull request:
In this PR I am addressing some comments in https://github.com/bitcoin/bitcoin/pull/27501 as a followup. 
- changed `prioritisation-map` in the `RPCResult` to `""`
- Directly constructing 2 entry map for getprioritisedtransactions in functional tests
- renamed `txid` to `transactionid` in `RPCResult` to be more consistent with naming elsewhere
- exposed the `modified_fee` field instead of having it be a useless arg
- Created a new test that asserts when `prioritisedtransaction` is called with a fee_delta of 0 it is not added to mempool

kevkevinpal committed:
test: Directly constructing 2 entry map for getprioritisedtransactions

Directly constructing the map in the assertion instead of indexing by txid
so that we wouldn't miss new entries in future regressions.

kevkevinpal committed:
rpc: changed prioritisation-map -> ""

prioritisation-map gets eaten by the help generator to be "" so we are
setting to "" to begin with

kevkevinpal committed:
test: Assert that a new tx with a delta of 0 is never added

luke-jr commented on src/rpc/mining.cpp:
```diff
@@ -507,7 +508,12 @@ static RPCHelpMan getprioritisedtransactions()
                 result_inner.pushKV("fee_delta", delta_info.delta);
+                CAmount modified_fee;
                 result_inner.pushKV("in_mempool", delta_info.in_mempool);
+                if(delta_info.in_mempool){
+                   modified_fee = *delta_info.modified_fee;
+                   result_inner.pushKV("modified_fee", (uint64_t)modified_fee);
+                }
```
This isn't a refactor.

kevkevinpal replied:
good point let me change the title

ajtowns commented:
Could you make the PR description a standalone statement of what this PR is trying to do, and leave the links to previous comments and questions in a separate comment?

kevkevinpal commented:
## Below are commits and corresponding comments they address

#### [80f7ccd](https://github.com/kevkevinpal/bitcoin/pull/6/commits/80f7ccdefda97cba9a7a913bf9c8ae19e368f7d9) test: Directly constructing 2 entry map for getprioritisedtransactions
- https://github.com/bitcoin/bitcoin/pull/27501#discussion_r1219668280
- https://github.com/bitcoin/bitcoin/pull/27501#discussion_r1219669146
- https://github.com/bitcoin/bitcoin/pull/27501#discussion_r1219669511

#### [be75730](https://github.com/kevkevinpal/bitcoin/pull/6/commits/be757305484815693eeef29a0858c8dd640b75c7) refactor: changed prioritisation-map -> ""
- https://github.com/bitcoin/bitcoin/pull/27501#discussion_r1220320231

#### [49332a7](https://github.com/kevkevinpal/bitcoin/pull/6/commits/49332a723aa1ca0b147f842d58797a1b9db24677) refactor: renaming txid -> transactionid
- https://github.com/bitcoin/bitcoin/pull/27501#discussion_r1220330124

#### [816ee54](https://github.com/kevkevinpal/bitcoin/pull/6/commits/816ee540a68a33bbf2528d833aca59c287171aa6) refactor: exposing modified_fee in getprioritisedtransactions
- https://github.com/bitcoin/bitcoin/pull/27501#discussion_r1191062084
- https://github.com/bitcoin/bitcoin/pull/27501#discussion_r1220349890

#### [7e54f44](https://github.com/kevkevinpal/bitcoin/pull/6/commits/7e54f4434f5391f26f9dce15025c800923a1e869) test: Assert that a new tx with a delta of 0 is never added
- https://github.com/bitcoin/bitcoin/pull/27501#discussion_r1219680499

#### TODO/comments not addressed yet in this PR

These might make more sense in their own pr?
https://github.com/bitcoin/bitcoin/pull/27501#discussion_r1184508118
https://github.com/bitcoin/bitcoin/pull/27501#discussion_r1186460872
https://github.com/bitcoin/bitcoin/pull/27501#discussion_r1220348795

Unsure if this is wanted
https://github.com/bitcoin/bitcoin/pull/27501#discussion_r1220309713

kevkevinpal commented on test/functional/mining_prioritisetransaction.py:
```diff
@@ -111,9 +111,8 @@ def test_diamond(self):
-        assert_equal(prioritisation_map_in_mempool[txid_b], {"fee_delta" : fee_delta_b*COIN, "in_mempool" : True})
-        assert_equal(prioritisation_map_in_mempool[txid_c], {"fee_delta" : (fee_delta_c_1 + fee_delta_c_2)*COIN, "in_mempool" : True})
+        assert_equal(self.nodes[0].getprioritisedtransactions()[txid_b], {"fee_delta" : fee_delta_b*COIN, "in_mempool" : True})
+        assert_equal(self.nodes[0].getprioritisedtransactions()[txid_c], {"fee_delta" : (fee_delta_c_1 + fee_delta_c_2)*COIN, "in_mempool" : True})
```
I think I may have miss read the [initial comment](https://github.com/bitcoin/bitcoin/pull/27501#discussion_r1219668280) but I have updated the message and the tests.

I am now testing the whole object we get back from `self.nodes[0].getprioritisedtransactions()` so in the future if there are new transactions added this assertion will fail

kevkevinpal commented:
looks like the "test each commit" git action failed but doesn't seems related to the changes proposed

I see the following in the logs
```
E: Failed to fetch mirror+file:/etc/apt/apt-mirrors.txt/pool/main/a/avahi/libavahi-core7_0.8-5ubuntu5.1_amd64.deb  404  Not Found [IP: 52.252.75.106 80]
E: Failed to fetch mirror+file:/etc/apt/apt-mirrors.txt/pool/main/a/avahi/avahi-daemon_0.8-5ubuntu5.1_amd64.deb  404  Not Found [IP: 52.252.75.106 80]
```

kevkevinpal commented on src/rpc/mining.cpp:
```diff
@@ -489,7 +489,7 @@ static RPCHelpMan getprioritisedtransactions()
             RPCResult::Type::OBJ_DYN, "", "prioritisation keyed by txid",
             {
-                {RPCResult::Type::OBJ, "txid", "", {
+                {RPCResult::Type::OBJ, "transactionid", "", {
```
updated the commit message in [9a918a4](https://github.com/bitcoin/bitcoin/pull/28885/commits/9a918a4720ee091344a16dcd703d169ebdb7938e) 

I also fixed the other ones that had refactor in the commit message since they weren't really refactors either

kevkevinpal commented on src/rpc/mining.cpp:
```diff
@@ -507,7 +508,12 @@ static RPCHelpMan getprioritisedtransactions()
                 result_inner.pushKV("in_mempool", delta_info.in_mempool);
+                if(delta_info.in_mempool){
+                   modified_fee = *delta_info.modified_fee;
+                   result_inner.pushKV("modified_fee", (uint64_t)modified_fee);
```
thanks! fixed in [e3f2fc0](https://github.com/bitcoin/bitcoin/pull/28885/commits/e3f2fc026a705148221de1bd0ecb37a26c3a4b36)

glozow commented on test/functional/mining_prioritisetransaction.py:
```diff
@@ -263,6 +263,12 @@ def run_test(self):
                 assert x not in mempool
 
+
+        self.log.info("Assert that prioritised free transaction with 0 delta is never added to mapDeltas")
+        tx_id_zero_del = self.wallet.create_self_transfer(fee_rate=0)['txid']
```
in a1fb3eddc91db708727fc48309d3bb6182858bbd

I don't think it matters what the feerate of this tx is?
```suggestion
        self.log.info("Assert that 0 delta is never added to mapDeltas")
        tx_id_zero_del = self.wallet.create_self_transfer()['txid']
```

kevkevinpal replied:
yes you are right updated the test in [0d3a4ab](https://github.com/bitcoin/bitcoin/pull/28885/commits/0d3a4abd805e3d4deec25f9902e762a7e693ce6b)

glozow commented on src/rpc/mining.cpp:
```diff
@@ -489,7 +489,7 @@ static RPCHelpMan getprioritisedtransactions()
             RPCResult::Type::OBJ_DYN, "", "prioritisation keyed by txid",
             {
-                {RPCResult::Type::OBJ, "txid", "", {
+                {RPCResult::Type::OBJ, "transactionid", "", {
```
9a918a4720ee091344a16dcd703d169ebdb7938e: Not sure if we should change results at this point?

kevkevinpal replied:
updated this in this force push da0036c97c4bf4ad8498ee8c27cdb395636aa6fb

glozow reviewed (commented):
Thanks for the followup!

kevkevinpal commented on src/rpc/mining.cpp:
```diff
@@ -508,6 +509,9 @@ static RPCHelpMan getprioritisedtransactions()
                 UniValue result_inner{UniValue::VOBJ};
                 result_inner.pushKV("fee_delta", delta_info.delta);
                 result_inner.pushKV("in_mempool", delta_info.in_mempool);
+                if(delta_info.in_mempool){
```
thanks! I didnt know this was a command I could use, I will save it for later. I applied it to my commits and force pushed

kevkevinpal commented on src/rpc/mining.cpp:
```diff
@@ -487,10 +487,11 @@ static RPCHelpMan getprioritisedtransactions()
-                {RPCResult::Type::OBJ, "txid", "", {
+                {RPCResult::Type::OBJ, "transactionid", "", {
                     {RPCResult::Type::NUM, "fee_delta", "transaction fee delta in satoshis"},
+                    {RPCResult::Type::NUM, "modified_fee", /*optional=*/true, "new fee for this transactionid if in mempool"},
```
makes sense thanks! force pushed in [2f11d8d](https://github.com/bitcoin/bitcoin/pull/28885/commits/2f11d8d07e9141c56f6f7cc4e406d73723ead6e3)

kevkevinpal commented on src/rpc/mining.cpp:
```diff
@@ -483,24 +483,21 @@ static RPCHelpMan prioritisetransaction()
-                    {RPCResult::Type::BOOL, "in_mempool", "whether this transaction is currently in mempool"},
-                }}
-            },
+            RPCResult::Type::OBJ_DYN,
```
@jonatack said that I should run the script in this [comment](https://github.com/bitcoin/bitcoin/pull/28885#discussion_r1406322235) and it lead to there being a indentation, I can remove if needed

glozow replied:
That's fine for lines that you're adding or already changing (which iiuc is what jonatack was suggesting as well). To make review easier, I wouldn't recommend reformatting other lines, as it crowds the diff.

kevkevinpal replied:
fixed in latest force push [fba275b](https://github.com/bitcoin/bitcoin/pull/28885/commits/fba275b39c83525e98540c110e09d5177fcdb4a0)

glozow commented:
Still working on this? I'll be available to re-review quickly after comments are addressed.

kevkevinpal commented:
> Still working on this? I'll be available to re-review quickly after comments are addressed.

yes sorry I can try and make some followup changes to the comments soon

kevkevinpal committed:
rpc: exposing modified_fee in getprioritisedtransactions

Instead of having modified_fee be hidden we are now exposing it to avoid
having useless code

kevkevinpal committed:
rpc: renaming txid -> transactionid

renamed to transactionid because it is named this way in getrawmempool
and getmempoolancestors

kevkevinpal commented on test/functional/mempool_expiry.py:
```diff
@@ -42,7 +42,8 @@ def test_transaction_expiry(self, timeout):
         node.prioritisetransaction(parent_txid, 0, COIN)
-        assert_equal(node.getprioritisedtransactions()[parent_txid], { "fee_delta" : COIN, "in_mempool" : True})
+        # Fee rate found in test/functional/test_framework/wallet.py create_self_transfer
+        assert_equal(node.getprioritisedtransactions()[parent_txid], { "fee_delta" : COIN, "in_mempool" : True, "modified_fee": COIN + COIN * (0.003 * 104 / 1000) })
```
Added a comment ontop showing where I got these numbers from

glozow replied:
These are still [magic numbers](https://en.wikipedia.org/wiki/Magic_number_(programming)), even if they are accurate. The comment is helpful, but doesn't change the fact that this test relies on the implementation of `create_self_transfer`, making it brittle. If we were to change the implementation so that the virtual size could be something other than 104, or if we adjusted the default feerate to something other than 0.003, this test would break, even though the mempool/RPC logic hasn't changed.

I was suggesting that you assign variables and reuse them in this test:
- `fee_delta` to use in `prioritisetransaction`
- `parent_fee` meaning the absolute fees of the tx. If you're having trouble getting the fee, it'll be one of the results returned from `send_self_transfer`.
- assert here that `modified_fee` is equal to `parent_fee + fee_delta`

kevkevinpal replied:
not sure why I didn't think of just using the value returned from `create_self_transfer` fixed in [24b490f](https://github.com/bitcoin/bitcoin/pull/28885/commits/24b490f292149e6151dcddaf11ac0d5bf16dde44)

kevkevinpal commented on test/functional/mining_prioritisetransaction.py:
```diff
@@ -45,15 +45,15 @@ def test_replacement(self):
         assert_equal(self.nodes[0].getprioritisedtransactions(), { tx_replacee["txid"] : { "fee_delta" : 100, "in_mempool" : False}})
         self.nodes[0].sendrawtransaction(tx_replacee["hex"])
-        assert_equal(self.nodes[0].getprioritisedtransactions(), { tx_replacee["txid"] : { "fee_delta" : 100, "in_mempool" : True}})
+        assert_equal(self.nodes[0].getprioritisedtransactions(), { tx_replacee["txid"] : { "fee_delta" : 100, "in_mempool" : True, "modified_fee": 1140}})
```
updated in 74fae193384b722462a9ede2d55b3375bad74332

kevkevinpal commented on test/functional/mining_prioritisetransaction.py:
```diff
@@ -263,6 +263,12 @@ def run_test(self):
+        self.log.info("Assert that 0 delta is never added to mapDeltas")
+        tx_id_zero_del = self.wallet.create_self_transfer()['txid']
+        self.nodes[0].prioritisetransaction(txid=tx_id_zero_del, fee_delta=0)
+        assert tx_id_zero_del not in self.nodes[0].getrawmempool()
```
yes you are right updated in [3fbcbc2](https://github.com/bitcoin/bitcoin/pull/28885/commits/3fbcbc26dd2ea0f3257ca52b3f5a037aad7b9b04)

glozow commented on src/rpc/mining.cpp:
```diff
@@ -492,6 +492,7 @@ static RPCHelpMan getprioritisedtransactions()
                 {RPCResult::Type::OBJ, "<transactionid>", "", {
                     {RPCResult::Type::NUM, "fee_delta", "transaction fee delta in satoshis"},
                     {RPCResult::Type::BOOL, "in_mempool", "whether this transaction is currently in mempool"},
+                    {RPCResult::Type::NUM, "modified_fee", /*optional=*/true, "new fee for this transaction. Only returned if transaction is in mempool"},
```
dc189d3ec5c4af9bca02928dfeb4770570060982: I'm not sure that "new fee" would be understood by people. It would also be helpful to add the units here, i.e.
```suggestion
                    {RPCResult::Type::NUM, "modified_fee", /*optional=*/true, "modified fee in satoshis. Only returned if in_mempool=true"},
```

kevkevinpal replied:
updated in cfdbcd19b32fd63954d7947dcc639aef291fb6b2

glozow reviewed (commented):
ACK 5353af1c8a2dd402a303a786d8aae5edbfc1ff3a

glozow commented:
reACK 0eebd6fe7d01ddc7f6b7f13a6ed6e705c7aeae4e, only change is the doc suggestion
